exchanges/huobi/huobi_websocket.go

```diff
@@ -548,284 +500,232 @@ func (h *HUOBI) generateSubscriptions() (subscription.List, error) {
-	ctx := context.Background()
-	var errs error
-	var creds *account.Credentials
-	if len(subs.Private()) > 0 {
-		if creds, errs = h.GetCredentials(ctx); errs != nil {
-			return errs
-		}
+	subs, errs := subs.ExpandTemplates(h)
+	return common.AppendError(errs, h.ParallelChanOp(subs, func(l subscription.List) error { return h.manageSubs(wsUnsubOp, l) }, 1))
+}
+
+func (h *HUOBI) manageSubs(op string, subs subscription.List) error {
+	if len(subs) != 1 {
+		return subscription.ErrBatchingNotSupported
 	}
-	for _, s := range subs {
-		var err error
-		if s.Authenticated {
-			err = h.wsAuthenticatedSubscribe(creds, "unsub", wsAccountsOrdersEndPoint+"/"+s.QualifiedChannel, s.QualifiedChannel)
+	s := subs[0]
+	var c stream.Connection
+	var req any
+	if s.Authenticated {
+		c = h.Websocket.AuthConn
+		req = wsReq{Action: op, Channel: s.QualifiedChannel}
+	} else {
+		c = h.Websocket.Conn
+		if op == wsSubOp {
+			// Set the id to the channel so that V1 errors can make it back to us
+			req = wsSubReq{ID: wsSubOp + ":" + s.QualifiedChannel, Sub: s.QualifiedChannel}
 		} else {
-			err = h.Websocket.Conn.SendJSONMessage(ctx, request.Unset, WsRequest{Unsubscribe: s.QualifiedChannel})
+			req = wsSubReq{Unsub: s.QualifiedChannel}
+		}
+	}
+	if op == wsSubOp {
+		s.SetKey(s.QualifiedChannel)
+		if err := h.Websocket.AddSubscriptions(c, s); err != nil {
+			return fmt.Errorf("%w: %s; error: %w", stream.ErrSubscriptionFailure, s, err)
 		}
-		if err == nil {
-			err = h.Websocket.RemoveSubscriptions(h.Websocket.Conn, s)
+	}
+	ctx := context.Background()
+	respRaw, err := c.SendMessageReturnResponse(ctx, request.Unset, wsSubOp+":"+s.QualifiedChannel, req)
+	if err == nil {
+		err = getErrResp(respRaw)
+	}
+	if err != nil {
+		if op == wsSubOp {
+			_ = h.Websocket.RemoveSubscriptions(c, s)
 		}
-		errs = common.AppendError(errs, err)
+		return fmt.Errorf("%s: %w", s, err)
 	}
-	return errs
+	if op == wsSubOp {
+		err = s.SetState(subscription.SubscribedState)
+		if h.Verbose {
+			log.Debugf(log.ExchangeSys, "%s Subscribed to %s", h.Name, s)
+		}
+	} else {
+		err = h.Websocket.RemoveSubscriptions(c, s)
+	}
+	return err
 }
 
-func (h *HUOBI) wsGenerateSignature(creds *account.Credentials, timestamp, endpoint string) ([]byte, error) {
+func (h *HUOBI) wsGenerateSignature(creds *account.Credentials, timestamp string) ([]byte, error) {
 	values := url.Values{}
-	values.Set("AccessKeyId", creds.Key)
-	values.Set("SignatureMethod", signatureMethod)
-	values.Set("SignatureVersion", signatureVersion)
-	values.Set("Timestamp", timestamp)
-	host := "api.huobi.pro"
-	payload := fmt.Sprintf("%s\n%s\n%s\n%s",
-		http.MethodGet, host, endpoint, values.Encode())
+	values.Set("accessKey", creds.Key)
+	values.Set("signatureMethod", signatureMethod)
+	values.Set("signatureVersion", signatureVersion)
+	values.Set("timestamp", timestamp)
+	payload := fmt.Sprintf("%s\n%s\n%s\n%s", http.MethodGet, wsSpotHost, wsPrivatePath, values.Encode())
```

Comment:
concatenate please instead, while you are here 🙏

Reply:
Fixed https://github.com/gbjk/gocryptotrader/commit/68f571f28